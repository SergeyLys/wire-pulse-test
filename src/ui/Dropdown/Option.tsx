import { memo } from "react";

type Props<T> = {
  onSelect?: (s: T) => void;
  value: T;
  default?: boolean;
  children: React.ReactNode;
}

type TestingProps = {
  testid?: string;
}

function Option<T>({ children, onSelect, value, testid }: Props<T> & TestingProps) {
  const handleClick = () => {
    if (onSelect) {
      onSelect(value);
    }
  };
  return (
    <div data-testid={testid || "dropdown-option"} className="dropdown-option" onClick={handleClick}>
      {children}
    </div>
  );
}

export default memo(Option);