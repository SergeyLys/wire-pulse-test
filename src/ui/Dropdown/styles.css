.dropdown {
  position: relative;
}

.dropdown-button {
  position: relative;
  background-color: #fff;
  color: #000;
  padding: 10px 22px 10px 10px;
  cursor: pointer;
  border-radius: 10px;
  border: 1px solid rgba(0, 0, 0, 0.5);
  box-shadow: 0 4px 6px rgba(0, 0, 0, 0.1);
  transition: box-shadow 0.3s ease;
  min-width: 65px;
  text-align: center;
}

.dropdown-button:hover {
  box-shadow: 0 4px 6px rgba(0, 0, 0, 0);
}

.dropdown-button:after {
  content: "\25BE";
  position: absolute;
  right: 10px;
  transform: rotate(0deg);
  transition: transform 0.3s ease;
}

.dropdown-button.open:after {
  transform: rotate(90deg);
}

.dropdown-content {
  position: absolute;
  top: 100%;
  left: 0;
  background-color: #fff;
  border: 1px solid #ccc;
  border-top: none;
  color: black;
  width: 100%;
}

.dropdown-option {
  padding: 10px;
  cursor: pointer;
  transition: background-color 0.3s ease;
}

.dropdown-option:hover {
  background-color: #f0f0f0;
}
