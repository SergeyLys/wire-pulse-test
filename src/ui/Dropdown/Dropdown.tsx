import React, {
  useEffect,
  useState,
  Children,
  useRef,
  useCallback,
  memo,
} from "react";
import Option from "./Option";
import "./styles.css";

type DropdownComposition = {
  Option: typeof Option;
}

type Props<T> = TestingProps & {
  onChange: (item: T) => void;
  children?: React.ReactElement | React.ReactElement[];
  placeholder?: string;
}

type TestingProps = {
  openerTestId?: string;
}

function Dropdown<T>({
  children,
  onChange,
  placeholder,
  openerTestId
}: Props<T>) {
  const [isOpen, setIsOpen] = useState<boolean>(false);
  const [selectedValue, setSelectedValue] = useState<string>(
    (
      Children.toArray(children).find(
        (child) => (child as React.ReactElement).props.default,
      ) as React.ReactElement
    )?.props.children || placeholder,
  );
  const containerRef = useRef<HTMLDivElement>(null);

  useEffect(() => {
    setSelectedValue(placeholder || selectedValue);
  }, [placeholder]);

  useEffect(() => {
    document.addEventListener("click", hideDropdown);
    return () => {
      document.removeEventListener("click", hideDropdown);
    };
  }, [isOpen]);

  const hideDropdown = (e: MouseEvent) => {
    if (e.target === containerRef.current || !isOpen)
      return;

    setIsOpen(false);
  };

  const toggleDropdown = useCallback(() => {
    setIsOpen(!isOpen);
  }, [isOpen]);

  return (
    <div className="dropdown">
      <div data-testid={openerTestId || "dropdown-trigger"} onClick={toggleDropdown}>
        <div
          ref={containerRef}
          className={`dropdown-button ${isOpen ? "open" : ""}`}
        >
          {selectedValue}
        </div>
      </div>
      {isOpen && (
        <div className="dropdown-content">
          {children && Children.map(children, (child) => {
            if (child.type === MemoizedDropdown.Option) {
              return React.cloneElement(child, {
                ...child.props,
                onSelect: (item: T) => {
                  toggleDropdown();
                  onChange(item);
                  setSelectedValue(child.props.children);
                },
              });
            }
            return null;
          })}
        </div>
      )}
    </div>
  );
}

//@ts-ignore
const MemoizedDropdown = memo(Dropdown) as React.NamedExoticComponent<Props<any>> & DropdownComposition;

MemoizedDropdown.Option = Option;

export type { Props };
export default MemoizedDropdown;
