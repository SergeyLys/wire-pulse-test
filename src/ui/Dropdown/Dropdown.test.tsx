import renderer from 'react-test-renderer';
import { render, screen, fireEvent } from '@testing-library/react';
import Dropdown, {Props} from './Dropdown';

const Component = (props: Props<any> & {shouldUseDefaultOption: boolean}) => (
  <Dropdown
    placeholder={props.placeholder}
    onChange={props.onChange}
  >
    <div data-testid="element-to-trim" />
    <Dropdown.Option default={props.shouldUseDefaultOption} value={10}>
      Option 1
    </Dropdown.Option>
  </Dropdown>
);

const setupComponent = (props?: Record<string, any>) => {
  const defaultProps = {
    onChange: jest.fn(),
    shouldUseDefaultOption: true
  };

  const mergedProps = {...defaultProps, ...props};

  return <Component {...mergedProps} />;
}

describe('Dropdown component', () => {

  it('renders correctly', () => {
    const tree = renderer.create(
      setupComponent()
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should render only related components', () => {
    render(setupComponent());

    expect(screen.queryByTestId('element-to-trim')).not.toBeInTheDocument();
  });

  it('should render placeholder if default option is not set', () => {
    const props = {
      shouldUseDefaultOption: false,
      placeholder: "Dropdown placeholder"
    };
    render(setupComponent(props));
    expect(screen.getByText('Dropdown placeholder')).toBeInTheDocument();
  });

  it('should trigger onChange fn on change', () => {
    const props = {
      onChange: jest.fn()
    }
    const mockOnChange = jest.spyOn(props, 'onChange');

    render(setupComponent(props));

    fireEvent.click(screen.getByTestId('dropdown-trigger'));

    expect(screen.getAllByTestId('dropdown-option')).toHaveLength(1);

    fireEvent.click(screen.getByTestId('dropdown-option'));

    expect(mockOnChange).toHaveBeenCalledWith(10);
  });

  it('should close by document click', () => {
    render(setupComponent());

    fireEvent.click(screen.getByTestId('dropdown-trigger'));

    expect(screen.getAllByTestId('dropdown-option')).toHaveLength(1);

    fireEvent.click(document);

    expect(screen.queryByTestId('dropdown-option')).not.toBeInTheDocument();
  });
});
