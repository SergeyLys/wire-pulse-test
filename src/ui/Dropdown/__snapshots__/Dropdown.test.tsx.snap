// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dropdown component renders correctly 1`] = `
<div
  className="dropdown"
>
  <div
    data-testid="dropdown-trigger"
    onClick={[Function]}
  >
    <div
      className="dropdown-button "
    >
      Option 1
    </div>
  </div>
</div>
`;
